internal/dot/graph_test.go

```diff
@@ -21,27 +21,170 @@
 package dot
 
 import (
+	"reflect"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 )
 
-func TestNodeString(t *testing.T) {
-	n1 := &Node{Type: "t1"}
-	n2 := &Node{Type: "t2", Name: "bar"}
-	n3 := &Node{Type: "t3", Group: "foo"}
+type t1 struct{}
+type t2 struct{}
+type t3 struct{}
 
-	assert.Equal(t, "t1", n1.String())
-	assert.Equal(t, "t2[name=bar]", n2.String())
-	assert.Equal(t, "t3[group=foo]", n3.String())
+var type1 = reflect.TypeOf(t1{})
+var type2 = reflect.TypeOf(t2{})
+var type3 = reflect.TypeOf(t3{})
+var type4 = reflect.TypeOf([]t3{})
+
+var n1 = &Node{Type: type1}
+var n2 = &Node{Type: type2, Name: "bar"}
+var n3 = &Node{Type: type3, Group: "foo"}
+var n4 = &Node{Type: type4, Group: "foo"}
+
+var p1 = &Param{Node: n1}
+var p2 = &Param{Node: n2}
+var p3 = &Param{Node: n4}
+
+var r1 = &Result{Node: n1}
+var r2 = &Result{Node: n2}
+var r3 = &Result{Node: n3, GroupIndex: 5}
+
+var g1 = &Group{Type: reflect.TypeOf(t1{}), Group: "group1"}
```

Comment:
I think it would be preferable to avoid these globals. It's okay to have duplication of test data for separate tests. We don't want a change here for one test to cause failures in multiple other tests because they all rely on the same test data.

Reply:
I wasn't sure of the best way to do it since there's a lot of repetitive test data needed, but what you said about failing other tests on change makes sense. I've moved them back into individual tests.